### DISCUSSION TITLE
How to obtain the dual multipliers with Gurobi and Casadi?

### DISCUSSION THREAD

**Discussion by lalajiayou** [casadi/casadi#3106] [Q&A]
Hi, everyone, I encountered the following problem:
For the same SOCP optimization problem that is with feasible solutions, gurobi can solve it, but cannot return the dual multiplier normally (NaN); On the contrary, Ipopt can both solve normally and return dual variables. So, my question is how to obtain the dual multipliers with Gurobi and Casadi?

The corresponding codes are as follows:

opti = casadi.Opti();
opti.solver('ipopt')
%opti = casadi.Opti('conic');
%opti.solver('gurobi')
[~,sol] = evalc('opti.solve();');
lambda = sol.value(opti.lam_g)

I have noticed the following discussion, but there are still some doubts. I am not sure which file to modify. If anyone could reply, I would greatly appreciate it!
https://groups.google.com/g/casadi-users/c/IWRTF1IcU3s/m/k45pn8leAAAJ

**Comment by jgillis** [comment]
Looks like this feature is indeed missing from our interface. I've created an issue

**Reply by lalajiayou** [reply]
Thank you very much, jgillis!
I wonder if there is any quick way to fix this problem, because I am currently using casadi+gurobi. The reason why I do not use IPOPT is because gurobi has higher solving efficiency for LP, QP and SOCP problems. Especially for large-scale optimization problems, the acceleration effect of gurobi, compared with Ipopt, is particularly significant.
By the way, I also tried cplex, but found that it also cannot return dual multipliers normally.
Thanks again! Have a nice day!

**Comment by jgillis** [comment]
Dual variables for linear constraints will be available in 3.6.1 soon

**Comment by ZhuZheart98** [comment]
Could you please describe how casadi uses gurobi?

> Thank you very much, jgillis!非常感谢你，吉利斯！ I wonder if there is any quick way to fix this problem, because I am currently using casadi+gurobi. The reason why I do not use IPOPT is because gurobi has higher solving efficiency for LP, QP and SOCP problems. Especially for large-scale optimization problems, the acceleration effect of gurobi, compared with Ipopt, is particularly significant.我想知道是否有任何快速解决此问题的方法，因为我目前正在使用casadi + gurobi。我之所以不使用IPOPT，是因为黑比对LP、QP和SOCP问题有更高的求解效率。特别是对于大规模优化问题，与 Ipopt 相比，gurobi 的加速效果尤为显著。 By the way, I also tried cplex, but found that it also cannot return dual multipliers normally.顺便说一句，我也尝试了 cplex，但发现它也无法正常返回双乘法器。 Thanks again! Have a nice day!再次感谢！有好的一天！
